In [1]:
import pandas as pd
import numpy as np
import matplotlib
from sklearn.metrics.pairwise import cosine_similarity
from scipy.optimize import minimize, rosen, rosen_der
from sklearn.model_selection import train_test_split

In [2]:
data = pd.read_csv("corpus/train.csv")
class_data = data[["antagonise", "condescending", "dismissive", "generalisation", "generalisation_unfair", "hostile", "sarcastic"]]
conf_data = data[['antagonise:confidence', 'condescending:confidence', 'dismissive:confidence','generalisation:confidence', 'generalisation_unfair',
       'generalisation_unfair:confidence','hostile:confidence', 'sarcastic:confidence']]
# np_data = np.array(data)
# print(data.columns)
# data_index = {name:i for i, name in enumerate(data.columns)}

# data_index["_unit_id"]
# class_data
# print(data)
conf_data

antagonise:confidence  condescending:confidence  dismissive:confidence  \
0                     0.5816                    0.5816                 0.5816   
1                     1.0000                    1.0000                 1.0000   
2                     1.0000                    0.8063                 1.0000   
3                     0.7931                    0.5959                 0.5959   
4                     1.0000                    1.0000                 1.0000   
...                      ...                       ...                    ...   
35498                 0.8114                    0.7888                 0.8114   
35499                 1.0000                    1.0000                 1.0000   
35500                 1.0000                    1.0000                 1.0000   
35501                 1.0000                    1.0000                 1.0000   
35502                 1.0000                    1.0000                 1.0000   

       generalisation:confidence  generalisation_unfair  \
0                         1.0000                    0.0   
1                         1.0000                    0.0   
2                         1.0000                    0.0   
3                         1.0000                    0.0   
4                         1.0000                    0.0   
...                          ...                    ...   
35498                     0.6002                    0.0   
35499                     1.0000                    0.0   
35500                     1.0000                    0.0   
35501                     1.0000                    0.0   
35502                     1.0000                    0.0   

       generalisation_unfair:confidence  hostile:confidence  \
0                                   1.0              0.5816   
1                                   1.0              1.0000   
2                                   1.0              1.0000   
3                                   1.0              1.0000   
4                                   1.0              1.0000   
...                                 ...                 ...   
35498                               0.0              0.6002   
35499                               1.0              1.0000   
35500                               1.0              1.0000   
35501                               1.0              1.0000   
35502                               1.0              1.0000   

       sarcastic:confidence  
0                    0.8001  
1                    1.0000  
2                    0.8063  
3                    0.6052  
4                    1.0000  
...                     ...  
35498                0.6002  
35499                1.0000  
35500                1.0000  
35501                1.0000  
35502                1.0000  

[35503 rows x 8 columns]

In [42]:
target_column = "healthy"
# data = data.drop(columns = ["comment", "_unit_id"])

xtrain, xtest, ytrain, ytest = train_test_split(class_data, data[target_column])

In [4]:
pos_sum = sum(data["healthy"])
print(f"""The total of comments is {len(data)}, of those {pos_sum} healthy, 
so saying everything is healthy would already result in a correct percentage of {pos_sum/ len(data)*100}%""")


The total of comments is 35503, of those 32848 healthy, 
so saying everything is healthy would already result in a correct percentage of 92.52175872461483%


In [5]:
# from sklearn.ensemble import RandomForestClassifier, VotingClassifier
# from sklearn.linear_model import LogisticRegression
# from sklearn.svm import SVC
# from sklearn.metrics import accuracy_score
# from sklearn.model_selection import train_test_split

# log_clf = LogisticRegression()
# rnd_clf = RandomForestClassifier(n_estimators=100)
# svm_clf = SVC()

# voting_clf = VotingClassifier(
#     estimators=[("lr", log_clf), ("rf", rnd_clf), ("svc", svm_clf)],
#     voting="hard",
# )
# voting_clf.fit(xtrain, ytrain)

# for clf in (log_clf, rnd_clf, svm_clf, voting_clf):
#     clf.fit(xtrain, ytrain)
#     y_pred = clf.predict(xtest)
#     print(clf.__class__.__name__, accuracy_score(ytest, y_pred))

In [6]:
from matplotlib.colors import ListedColormap
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.datasets import make_moons, make_circles, make_classification
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier, VotingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis

def classify(xtrain, xtest,ytrain, ytest, voting="hard"):
    # Add all the estemators/classifiers
    est = [("lr", LogisticRegression()), 
           ("rf", RandomForestClassifier(n_estimators=100)), 
           ("nb", GaussianNB()),
            ("qda", QuadraticDiscriminantAnalysis()),
           ("kn", KNeighborsClassifier()),
           ("mlp", MLPClassifier())
           ]
    voting_clf = VotingClassifier(
        estimators=est,
        voting=voting,
    )
    voting_clf.fit(xtrain, ytrain)

    for _, clf in est:
        clf.fit(xtrain, ytrain)
        y_pred = clf.predict(xtest)
        print(clf.__class__.__name__, accuracy_score(ytest, y_pred))

    # And now we print the voting clf
    voting_clf.fit(xtrain, ytrain)
    y_pred = voting_clf.predict(xtest)
    print(voting_clf.__class__.__name__, accuracy_score(ytest, y_pred))
classify(xtrain, xtest,ytrain, ytest)

LogisticRegression 0.9441189725101397
RandomForestClassifier 0.9450202794051374
GaussianNB 0.9085173501577287
QuadraticDiscriminantAnalysis 0.9085173501577287
KNeighborsClassifier 0.9433303289770166
MLPClassifier 0.9468228931951329
VotingClassifier 0.9453582694907616


## Soft voting

In [7]:
classify(xtrain, xtest,ytrain, ytest, "soft")

LogisticRegression 0.9441189725101397
RandomForestClassifier 0.9454709328526363
GaussianNB 0.9085173501577287
QuadraticDiscriminantAnalysis 0.9085173501577287
KNeighborsClassifier 0.9433303289770166
MLPClassifier 0.9469355565570077
VotingClassifier 0.9423163587201442


# Confidence attempt

In [8]:
np.array(conf_data)
weighted = []
for conf_column, class_column in zip(conf_data, class_data):
    column = []
    for conf, clas in zip(data[conf_column], data[class_column]):
        if clas == 1:
            new = conf
        else: 
            new = 1 - conf
        column.append(new)
    weighted.append(column)
    
    
weightedold = weighted    
weighted = np.array(weighted).T
print(weighted)




[[0.5816 0.5816 0.5816 ... 1.     1.     0.4184]
 [0.     0.     0.     ... 1.     0.     0.    ]
 [0.     0.1937 0.     ... 1.     0.     0.    ]
 ...
 [0.     0.     0.     ... 1.     0.     0.    ]
 [0.     0.     0.     ... 1.     0.     0.    ]
 [0.     0.     0.     ... 1.     0.     0.    ]]


In [9]:
xtrain, xtest, ytrain, ytest = train_test_split(weighted, data[target_column])
classify(xtrain, xtest,ytrain, ytest)

C:\Users\timho\anaconda3\lib\site-packages\sklearn\discriminant_analysis.py:715: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")


LogisticRegression 0.9426543488057684
RandomForestClassifier 0.9427670121676431
GaussianNB 0.893195132942767
QuadraticDiscriminantAnalysis 0.7198062190175755


C:\Users\timho\anaconda3\lib\site-packages\sklearn\discriminant_analysis.py:715: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")


KNeighborsClassifier 0.9400630914826499
MLPClassifier 0.9459215863001352


C:\Users\timho\anaconda3\lib\site-packages\sklearn\discriminant_analysis.py:715: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")


VotingClassifier 0.9410770617395223


In [10]:
classify(xtrain, xtest,ytrain, ytest, "soft")

C:\Users\timho\anaconda3\lib\site-packages\sklearn\discriminant_analysis.py:715: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")


LogisticRegression 0.9426543488057684
RandomForestClassifier 0.9436683190626408
GaussianNB 0.893195132942767
QuadraticDiscriminantAnalysis 0.7198062190175755


C:\Users\timho\anaconda3\lib\site-packages\sklearn\discriminant_analysis.py:715: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")


KNeighborsClassifier 0.9400630914826499
MLPClassifier 0.9452456061288869


C:\Users\timho\anaconda3\lib\site-packages\sklearn\discriminant_analysis.py:715: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")


VotingClassifier 0.9349932401982876


In [11]:
# import seaborn as sns
# correlations = pd.DataFrame(weightedold).corr()
# sns.heatmap(correlations)


In [12]:
from sklearn.ensemble import BaggingClassifier

from sklearn.tree import DecisionTreeClassifier
bag_clf = BaggingClassifier(
    DecisionTreeClassifier(splitter="random", max_leaf_nodes=16),
    n_estimators=500,
    max_samples=1.0,
    bootstrap=True,
)
bag_clf.fit(xtrain, ytrain)
y_pred = bag_clf.predict(xtest)

bag_clf.fit(xtrain, ytrain)
y_pred = bag_clf.predict(xtest)
print(bag_clf.__class__.__name__, accuracy_score(ytest, y_pred))

BaggingClassifier 0.945583596214511


In [17]:
from keras.models import Sequential
from keras.layers import Dense
# load the dataset

# define the keras model
model = Sequential()
model.add(Dense(12, input_dim=7, activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(1, activation='sigmoid'))
# compile the keras model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
# fit the keras model on the dataset
model.fit(xtrain, ytrain, epochs=150, batch_size=10)
# evaluate the keras model
_, accuracy = model.evaluate(xtrain, ytrain)
print('Accuracy: %.2f' % (accuracy*100))

Epoch 1/150
2663/2663 [==============================] - 1s 436us/step - loss: 0.1868 - accuracy: 0.9380
Epoch 2/150
2663/2663 [==============================] - 1s 424us/step - loss: 0.1578 - accuracy: 0.9443
Epoch 3/150
2663/2663 [==============================] - 1s 425us/step - loss: 0.1570 - accuracy: 0.9455
Epoch 4/150
2663/2663 [==============================] - 1s 425us/step - loss: 0.1565 - accuracy: 0.9451
Epoch 5/150
2663/2663 [==============================] - 1s 425us/step - loss: 0.1557 - accuracy: 0.9463
Epoch 6/150
2663/2663 [==============================] - 1s 421us/step - loss: 0.1551 - accuracy: 0.9470
Epoch 7/150
2663/2663 [==============================] - 1s 422us/step - loss: 0.1548 - accuracy: 0.9469
Epoch 8/150
2663/2663 [==============================] - 1s 433us/step - loss: 0.1545 - accuracy: 0.9460
Epoch 9/150
2663/2663 [==============================] - 1s 430us/step - loss: 0.1543 - accuracy: 0.9467
Epoch 10/150
2663/2663 [==============================]

2663/2663 [==============================] - 1s 444us/step - loss: 0.1518 - accuracy: 0.9475
Epoch 79/150
2663/2663 [==============================] - 1s 431us/step - loss: 0.1516 - accuracy: 0.9479
Epoch 80/150
2663/2663 [==============================] - 1s 445us/step - loss: 0.1519 - accuracy: 0.9472
Epoch 81/150
2663/2663 [==============================] - 1s 422us/step - loss: 0.1517 - accuracy: 0.9470
Epoch 82/150
2663/2663 [==============================] - 1s 424us/step - loss: 0.1520 - accuracy: 0.9472
Epoch 83/150
2663/2663 [==============================] - 1s 423us/step - loss: 0.1519 - accuracy: 0.9476
Epoch 84/150
2663/2663 [==============================] - 1s 424us/step - loss: 0.1519 - accuracy: 0.9472
Epoch 85/150
2663/2663 [==============================] - 1s 428us/step - loss: 0.1517 - accuracy: 0.9476
Epoch 86/150
2663/2663 [==============================] - 1s 431us/step - loss: 0.1520 - accuracy: 0.9469
Epoch 87/150
2663/2663 [==============================] - 1

In [18]:
# define the keras model
model = Sequential()
model.add(Dense(12, input_dim=7, activation='sigmoid'))
model.add(Dense(8, activation='sigmoid'))
model.add(Dense(1, activation='sigmoid'))
# compile the keras model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
# fit the keras model on the dataset
model.fit(xtrain, ytrain, epochs=250, batch_size=10)
# evaluate the keras model
_, accuracy = model.evaluate(xtrain, ytrain)
print('Accuracy: %.2f' % (accuracy*100))

Epoch 1/250
2663/2663 [==============================] - 1s 434us/step - loss: 0.2247 - accuracy: 0.9285
Epoch 2/250
2663/2663 [==============================] - 1s 434us/step - loss: 0.1651 - accuracy: 0.9426
Epoch 3/250
2663/2663 [==============================] - 1s 434us/step - loss: 0.1629 - accuracy: 0.9423
Epoch 4/250
2663/2663 [==============================] - 1s 443us/step - loss: 0.1617 - accuracy: 0.9437
Epoch 5/250
2663/2663 [==============================] - 1s 435us/step - loss: 0.1609 - accuracy: 0.9439
Epoch 6/250
2663/2663 [==============================] - 1s 430us/step - loss: 0.1602 - accuracy: 0.9444
Epoch 7/250
2663/2663 [==============================] - 1s 433us/step - loss: 0.1594 - accuracy: 0.9442
Epoch 8/250
2663/2663 [==============================] - 1s 430us/step - loss: 0.1594 - accuracy: 0.9445
Epoch 9/250
2663/2663 [==============================] - 1s 432us/step - loss: 0.1592 - accuracy: 0.9450
Epoch 10/250
2663/2663 [==============================]

2663/2663 [==============================] - 1s 413us/step - loss: 0.1553 - accuracy: 0.9457
Epoch 79/250
2663/2663 [==============================] - 1s 417us/step - loss: 0.1551 - accuracy: 0.9461
Epoch 80/250
2663/2663 [==============================] - 1s 425us/step - loss: 0.1552 - accuracy: 0.9456
Epoch 81/250
2663/2663 [==============================] - 1s 433us/step - loss: 0.1551 - accuracy: 0.9452
Epoch 82/250
2663/2663 [==============================] - 1s 418us/step - loss: 0.1553 - accuracy: 0.9452
Epoch 83/250
2663/2663 [==============================] - 1s 423us/step - loss: 0.1552 - accuracy: 0.9452
Epoch 84/250
2663/2663 [==============================] - 1s 420us/step - loss: 0.1551 - accuracy: 0.9456
Epoch 85/250
2663/2663 [==============================] - 1s 425us/step - loss: 0.1552 - accuracy: 0.9455
Epoch 86/250
2663/2663 [==============================] - 1s 434us/step - loss: 0.1552 - accuracy: 0.9460
Epoch 87/250
2663/2663 [==============================] - 1

2663/2663 [==============================] - 1s 416us/step - loss: 0.1535 - accuracy: 0.9467
Epoch 155/250
2663/2663 [==============================] - 1s 422us/step - loss: 0.1537 - accuracy: 0.9467
Epoch 156/250
2663/2663 [==============================] - 1s 424us/step - loss: 0.1535 - accuracy: 0.9466
Epoch 157/250
2663/2663 [==============================] - 1s 423us/step - loss: 0.1535 - accuracy: 0.9470
Epoch 158/250
2663/2663 [==============================] - 1s 429us/step - loss: 0.1534 - accuracy: 0.9470
Epoch 159/250
2663/2663 [==============================] - 1s 426us/step - loss: 0.1534 - accuracy: 0.9470
Epoch 160/250
2663/2663 [==============================] - 1s 439us/step - loss: 0.1534 - accuracy: 0.9470
Epoch 161/250
2663/2663 [==============================] - 1s 421us/step - loss: 0.1535 - accuracy: 0.9470
Epoch 162/250
2663/2663 [==============================] - 1s 432us/step - loss: 0.1534 - accuracy: 0.9463
Epoch 163/250
2663/2663 [==========================

2663/2663 [==============================] - 1s 428us/step - loss: 0.1526 - accuracy: 0.9473
Epoch 231/250
2663/2663 [==============================] - 1s 431us/step - loss: 0.1526 - accuracy: 0.9472
Epoch 232/250
2663/2663 [==============================] - 1s 428us/step - loss: 0.1528 - accuracy: 0.9467
Epoch 233/250
2663/2663 [==============================] - 1s 425us/step - loss: 0.1527 - accuracy: 0.9468
Epoch 234/250
2663/2663 [==============================] - 1s 430us/step - loss: 0.1525 - accuracy: 0.9470
Epoch 235/250
2663/2663 [==============================] - 1s 433us/step - loss: 0.1527 - accuracy: 0.9468
Epoch 236/250
2663/2663 [==============================] - 1s 429us/step - loss: 0.1527 - accuracy: 0.9470
Epoch 237/250
2663/2663 [==============================] - 1s 429us/step - loss: 0.1528 - accuracy: 0.9468
Epoch 238/250
2663/2663 [==============================] - 1s 431us/step - loss: 0.1527 - accuracy: 0.9468
Epoch 239/250
2663/2663 [==========================

In [36]:
# define the keras model
model = Sequential()
model.add(Dense(24, input_dim=15, activation='relu'))
model.add(Dense(16, activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(1, activation='sigmoid'))
# compile the keras model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
# fit the keras model on the dataset
model.fit(xtrain, ytrain, epochs=250, batch_size=10)
# evaluate the keras model
_, accuracy = model.evaluate(xtest, ytest)
print('Accuracy: %.2f' % (accuracy*100))

Epoch 1/250
2663/2663 [==============================] - 1s 435us/step - loss: 0.1753 - accuracy: 0.9427
Epoch 2/250
2663/2663 [==============================] - 1s 459us/step - loss: 0.1612 - accuracy: 0.9457
Epoch 3/250
2663/2663 [==============================] - 1s 444us/step - loss: 0.1582 - accuracy: 0.9459
Epoch 4/250
2663/2663 [==============================] - 1s 434us/step - loss: 0.1574 - accuracy: 0.9462
Epoch 5/250
2663/2663 [==============================] - 1s 434us/step - loss: 0.1568 - accuracy: 0.9464
Epoch 6/250
2663/2663 [==============================] - 1s 427us/step - loss: 0.1561 - accuracy: 0.9458
Epoch 7/250
2663/2663 [==============================] - 1s 432us/step - loss: 0.1559 - accuracy: 0.9460
Epoch 8/250
2663/2663 [==============================] - 1s 419us/step - loss: 0.1556 - accuracy: 0.9467
Epoch 9/250
2663/2663 [==============================] - 1s 418us/step - loss: 0.1554 - accuracy: 0.9464
Epoch 10/250
2663/2663 [==============================]

2663/2663 [==============================] - 1s 435us/step - loss: 0.1472 - accuracy: 0.9487
Epoch 79/250
2663/2663 [==============================] - 1s 447us/step - loss: 0.1471 - accuracy: 0.9490
Epoch 80/250
2663/2663 [==============================] - 1s 434us/step - loss: 0.1468 - accuracy: 0.9493
Epoch 81/250
2663/2663 [==============================] - 1s 431us/step - loss: 0.1470 - accuracy: 0.9493
Epoch 82/250
2663/2663 [==============================] - 1s 434us/step - loss: 0.1466 - accuracy: 0.9497
Epoch 83/250
2663/2663 [==============================] - 1s 430us/step - loss: 0.1468 - accuracy: 0.9496
Epoch 84/250
2663/2663 [==============================] - 1s 427us/step - loss: 0.1471 - accuracy: 0.9498
Epoch 85/250
2663/2663 [==============================] - 1s 451us/step - loss: 0.1466 - accuracy: 0.9497
Epoch 86/250
2663/2663 [==============================] - 1s 431us/step - loss: 0.1469 - accuracy: 0.9488
Epoch 87/250
2663/2663 [==============================] - 1

2663/2663 [==============================] - 1s 496us/step - loss: 0.1433 - accuracy: 0.9507
Epoch 155/250
2663/2663 [==============================] - 1s 458us/step - loss: 0.1430 - accuracy: 0.9506
Epoch 156/250
2663/2663 [==============================] - 1s 461us/step - loss: 0.1432 - accuracy: 0.9504
Epoch 157/250
2663/2663 [==============================] - 1s 498us/step - loss: 0.1435 - accuracy: 0.9507
Epoch 158/250
2663/2663 [==============================] - 1s 467us/step - loss: 0.1428 - accuracy: 0.9505
Epoch 159/250
2663/2663 [==============================] - 1s 537us/step - loss: 0.1426 - accuracy: 0.9505
Epoch 160/250
2663/2663 [==============================] - 2s 584us/step - loss: 0.1428 - accuracy: 0.9517
Epoch 161/250
2663/2663 [==============================] - 1s 521us/step - loss: 0.1425 - accuracy: 0.9506
Epoch 162/250
2663/2663 [==============================] - 1s 542us/step - loss: 0.1425 - accuracy: 0.9508
Epoch 163/250
2663/2663 [==========================

2663/2663 [==============================] - 1s 505us/step - loss: 0.1405 - accuracy: 0.9512
Epoch 231/250
2663/2663 [==============================] - 1s 471us/step - loss: 0.1407 - accuracy: 0.9520
Epoch 232/250
2663/2663 [==============================] - 1s 454us/step - loss: 0.1404 - accuracy: 0.9517
Epoch 233/250
2663/2663 [==============================] - 1s 451us/step - loss: 0.1401 - accuracy: 0.9519
Epoch 234/250
2663/2663 [==============================] - 1s 448us/step - loss: 0.1408 - accuracy: 0.9513
Epoch 235/250
2663/2663 [==============================] - 1s 491us/step - loss: 0.1400 - accuracy: 0.9515
Epoch 236/250
2663/2663 [==============================] - 1s 468us/step - loss: 0.1401 - accuracy: 0.9519
Epoch 237/250
2663/2663 [==============================] - 1s 443us/step - loss: 0.1401 - accuracy: 0.9511
Epoch 238/250
2663/2663 [==============================] - 1s 450us/step - loss: 0.1403 - accuracy: 0.9517
Epoch 239/250
2663/2663 [==========================

# With a keras model, we've aquired an accuracy of 95.15%

In [25]:
# define the keras model
model = Sequential()
model.add(Dense(60, input_dim=7, activation='relu'))
for node_numb in [50,45,40,35,30,28,25,22,20,18,15,12,10,8,5]:
    model.add(Dense(node_numb, activation='relu'))
model.add(Dense(1, activation='sigmoid'))
# compile the keras model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
# fit the keras model on the dataset
model.fit(xtrain, ytrain, epochs=250, batch_size=10)
# evaluate the keras model
_, accuracy = model.evaluate(xtest,ytest)
print('Accuracy: %.2f' % (accuracy*100))

Epoch 1/250
2663/2663 [==============================] - 2s 762us/step - loss: 0.1820 - accuracy: 0.9340
Epoch 2/250
2663/2663 [==============================] - 2s 749us/step - loss: 0.1629 - accuracy: 0.9431
Epoch 3/250
2663/2663 [==============================] - 2s 756us/step - loss: 0.1606 - accuracy: 0.9454
Epoch 4/250
2663/2663 [==============================] - 2s 752us/step - loss: 0.1601 - accuracy: 0.9456
Epoch 5/250
2663/2663 [==============================] - 2s 767us/step - loss: 0.1596 - accuracy: 0.9454
Epoch 6/250
2663/2663 [==============================] - 2s 761us/step - loss: 0.1582 - accuracy: 0.9462
Epoch 7/250
2663/2663 [==============================] - 2s 758us/step - loss: 0.1570 - accuracy: 0.9461
Epoch 8/250
2663/2663 [==============================] - 2s 758us/step - loss: 0.1565 - accuracy: 0.9463
Epoch 9/250
2663/2663 [==============================] - 2s 751us/step - loss: 0.1570 - accuracy: 0.9463
Epoch 10/250
2663/2663 [==============================]

2663/2663 [==============================] - 2s 800us/step - loss: 0.1474 - accuracy: 0.9493
Epoch 79/250
2663/2663 [==============================] - 2s 774us/step - loss: 0.1477 - accuracy: 0.9475
Epoch 80/250
2663/2663 [==============================] - 2s 781us/step - loss: 0.1477 - accuracy: 0.9485
Epoch 81/250
2663/2663 [==============================] - 2s 764us/step - loss: 0.1468 - accuracy: 0.9485
Epoch 82/250
2663/2663 [==============================] - 2s 759us/step - loss: 0.1476 - accuracy: 0.9490
Epoch 83/250
2663/2663 [==============================] - 2s 749us/step - loss: 0.1470 - accuracy: 0.9491
Epoch 84/250
2663/2663 [==============================] - 2s 744us/step - loss: 0.1477 - accuracy: 0.9487
Epoch 85/250
2663/2663 [==============================] - 2s 745us/step - loss: 0.1468 - accuracy: 0.9490
Epoch 86/250
2663/2663 [==============================] - 2s 750us/step - loss: 0.1469 - accuracy: 0.9490
Epoch 87/250
2663/2663 [==============================] - 2

1843/2663 [===================>..........] - ETA: 0s - loss: 0.1419 - accuracy: 0.9505

KeyboardInterrupt: 

In [39]:
# define the keras model
model = Sequential()
model.add(Dense(60, input_dim=15, activation='relu'))
for node_numb in [50,45,40,35,30,28,25,22,20,18,15,12,10,8,5]:
    model.add(Dense(node_numb, activation='relu'))
model.add(Dense(1, activation='sigmoid'))
# compile the keras model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
# fit the keras model on the dataset
model.fit(xtrain, ytrain, epochs=150, batch_size=100)
# evaluate the keras model
_, accuracy = model.evaluate(xtest,ytest)
print('Accuracy: %.2f' % (accuracy*100))

Epoch 1/150
267/267 [==============================] - 0s 952us/step - loss: 0.2333 - accuracy: 0.9297
Epoch 2/150
267/267 [==============================] - 0s 1ms/step - loss: 0.1640 - accuracy: 0.9457
Epoch 3/150
267/267 [==============================] - 0s 1ms/step - loss: 0.1580 - accuracy: 0.9455
Epoch 4/150
267/267 [==============================] - 0s 1ms/step - loss: 0.1569 - accuracy: 0.9460
Epoch 5/150
267/267 [==============================] - 0s 1ms/step - loss: 0.1564 - accuracy: 0.9458
Epoch 6/150
267/267 [==============================] - 0s 1ms/step - loss: 0.1561 - accuracy: 0.9458
Epoch 7/150
267/267 [==============================] - 0s 1ms/step - loss: 0.1546 - accuracy: 0.9461
Epoch 8/150
267/267 [==============================] - 0s 945us/step - loss: 0.1543 - accuracy: 0.9461
Epoch 9/150
267/267 [==============================] - 0s 982us/step - loss: 0.1533 - accuracy: 0.9461
Epoch 10/150
267/267 [==============================] - 0s 963us/step - loss: 0.1542 

267/267 [==============================] - 0s 1ms/step - loss: 0.1419 - accuracy: 0.9518
Epoch 81/150
267/267 [==============================] - 0s 963us/step - loss: 0.1418 - accuracy: 0.9515
Epoch 82/150
267/267 [==============================] - 0s 963us/step - loss: 0.1421 - accuracy: 0.9517
Epoch 83/150
267/267 [==============================] - 0s 945us/step - loss: 0.1414 - accuracy: 0.9519
Epoch 84/150
267/267 [==============================] - 0s 971us/step - loss: 0.1417 - accuracy: 0.9515
Epoch 85/150
267/267 [==============================] - 0s 941us/step - loss: 0.1419 - accuracy: 0.9511
Epoch 86/150
267/267 [==============================] - 0s 948us/step - loss: 0.1419 - accuracy: 0.9512
Epoch 87/150
267/267 [==============================] - 0s 945us/step - loss: 0.1417 - accuracy: 0.9515
Epoch 88/150
267/267 [==============================] - 0s 1ms/step - loss: 0.1417 - accuracy: 0.9520
Epoch 89/150
267/267 [==============================] - 0s 933us/step - loss: 0.1

In [45]:
# define the keras model
model = Sequential()
model.add(Dense(12, input_dim=7, activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(1, activation='sigmoid'))
# compile the keras model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
# fit the keras model on the dataset
model.fit(xtrain, ytrain, epochs=150, batch_size=2000)
# evaluate the keras model
_, accuracy = model.evaluate(xtest,ytest)
print('Accuracy: %.2f' % (accuracy*100))

Epoch 1/150
14/14 [==============================] - 0s 858us/step - loss: 0.6809 - accuracy: 0.8575
Epoch 2/150
14/14 [==============================] - 0s 786us/step - loss: 0.6503 - accuracy: 0.9211
Epoch 3/150
14/14 [==============================] - 0s 1ms/step - loss: 0.6161 - accuracy: 0.9257
Epoch 4/150
14/14 [==============================] - 0s 858us/step - loss: 0.5778 - accuracy: 0.9246
Epoch 5/150
14/14 [==============================] - 0s 858us/step - loss: 0.5358 - accuracy: 0.9243
Epoch 6/150
14/14 [==============================] - 0s 1ms/step - loss: 0.4909 - accuracy: 0.9242
Epoch 7/150
14/14 [==============================] - 0s 1ms/step - loss: 0.4451 - accuracy: 0.9242
Epoch 8/150
14/14 [==============================] - 0s 858us/step - loss: 0.3998 - accuracy: 0.9244
Epoch 9/150
14/14 [==============================] - 0s 1ms/step - loss: 0.3577 - accuracy: 0.9271
Epoch 10/150
14/14 [==============================] - 0s 858us/step - loss: 0.3199 - accuracy: 0.93

14/14 [==============================] - 0s 786us/step - loss: 0.1756 - accuracy: 0.9455
Epoch 83/150
14/14 [==============================] - 0s 858us/step - loss: 0.1756 - accuracy: 0.9455
Epoch 84/150
14/14 [==============================] - 0s 858us/step - loss: 0.1756 - accuracy: 0.9455
Epoch 85/150
14/14 [==============================] - 0s 858us/step - loss: 0.1755 - accuracy: 0.9455
Epoch 86/150
14/14 [==============================] - 0s 929us/step - loss: 0.1755 - accuracy: 0.9456
Epoch 87/150
14/14 [==============================] - 0s 929us/step - loss: 0.1755 - accuracy: 0.9456
Epoch 88/150
14/14 [==============================] - 0s 1ms/step - loss: 0.1755 - accuracy: 0.9456
Epoch 89/150
14/14 [==============================] - 0s 858us/step - loss: 0.1755 - accuracy: 0.9456
Epoch 90/150
14/14 [==============================] - 0s 929us/step - loss: 0.1755 - accuracy: 0.9456
Epoch 91/150
14/14 [==============================] - 0s 858us/step - loss: 0.1755 - accuracy: 0.

In [ ]:
#target_column = "healthy"
# data = data.drop(columns = ["comment", "_unit_id"])
#all_data = data[["antagonise", "condescending", "dismissive", "generalisation", "generalisation_unfair", "hostile", 
#                   "sarcastic",'antagonise:confidence', 'condescending:confidence', 'dismissive:confidence',
#                   'generalisation:confidence', 'generalisation_unfair', 'generalisation_unfair:confidence',
#                   'hostile:confidence', 'sarcastic:confidence']]

In [34]:


xtrain, xtest, ytrain, ytest = train_test_split(all_data, data[target_column])
# define the keras model
model = Sequential()
model.add(Dense(60, input_dim=15, activation='relu'))
for node_numb in [50,45,40,35,30,28,25,22,20,18,15,12,10,8,5]:
    model.add(Dense(node_numb, activation='relu'))
model.add(Dense(1, activation='sigmoid'))
# compile the keras model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
# fit the keras model on the dataset
model.fit(xtrain, ytrain, epochs=150, batch_size=20)
# evaluate the keras model
_, accuracy = model.evaluate(xtest,ytest)
print('Accuracy: %.2f' % (accuracy*100))

Epoch 1/150
1332/1332 [==============================] - 1s 760us/step - loss: 0.1887 - accuracy: 0.9380
Epoch 2/150
1332/1332 [==============================] - 1s 765us/step - loss: 0.1667 - accuracy: 0.9443
Epoch 3/150
1332/1332 [==============================] - 1s 758us/step - loss: 0.1622 - accuracy: 0.9454
Epoch 4/150
1332/1332 [==============================] - 1s 766us/step - loss: 0.1623 - accuracy: 0.9458
Epoch 5/150
1332/1332 [==============================] - 1s 761us/step - loss: 0.1603 - accuracy: 0.9456
Epoch 6/150
1332/1332 [==============================] - 1s 762us/step - loss: 0.1595 - accuracy: 0.9453
Epoch 7/150
1332/1332 [==============================] - 1s 771us/step - loss: 0.1566 - accuracy: 0.9463
Epoch 8/150
1332/1332 [==============================] - 1s 819us/step - loss: 0.1566 - accuracy: 0.9463
Epoch 9/150
1332/1332 [==============================] - 1s 829us/step - loss: 0.1572 - accuracy: 0.9460
Epoch 10/150
1332/1332 [==============================]

1332/1332 [==============================] - 1s 770us/step - loss: 0.1464 - accuracy: 0.9492
Epoch 79/150
1332/1332 [==============================] - 1s 778us/step - loss: 0.1467 - accuracy: 0.9500
Epoch 80/150
1332/1332 [==============================] - 1s 769us/step - loss: 0.1465 - accuracy: 0.9496
Epoch 81/150
1332/1332 [==============================] - 1s 770us/step - loss: 0.1463 - accuracy: 0.9495
Epoch 82/150
1332/1332 [==============================] - 1s 772us/step - loss: 0.1463 - accuracy: 0.9490
Epoch 83/150
1332/1332 [==============================] - 1s 766us/step - loss: 0.1455 - accuracy: 0.9496
Epoch 84/150
1332/1332 [==============================] - 1s 772us/step - loss: 0.1456 - accuracy: 0.9502
Epoch 85/150
1332/1332 [==============================] - 1s 781us/step - loss: 0.1472 - accuracy: 0.9488
Epoch 86/150
1332/1332 [==============================] - 1s 769us/step - loss: 0.1459 - accuracy: 0.9500
Epoch 87/150
1332/1332 [==============================] - 1